src/main/java/io/dgraph/DgraphClient.java

```diff
@@ -31,6 +41,154 @@ public class DgraphClient {
+   * username and password are provided.
+   * <p>The connection string has the format: "dgraph://[username:password@]host:port[?params]"
+   * <p>Supported query parameters:
+   * <ul>
+   *   <li>sslmode - SSL connection mode. Supported values:
+   *     <ul>
+   *       <li>"disable" - No encryption, uses plaintext</li>
+   *       <li>"require" - Uses TLS encryption without certificate verification</li>
+   *       <li>"verify-ca" - Uses TLS encryption with certificate verification</li>
+   *     </ul>
+   *   </li>
+   *   <li>apikey - API key for authorization from Dgraph Cloud</li>
+   *   <li>bearertoken - Bearer token for authorization</li>
+   * </ul>
+   *
+   * @param connectionString The connection string to connect to Dgraph
+   * @return A new DgraphClient instance
+   * @throws IllegalArgumentException If the connection string is invalid
+   * @throws MalformedURLException If the connection string cannot be parsed as a URL
+   * @throws SSLException If there's an error configuring the SSL context for sslmode=require
+   */
+  public static DgraphClient openConnection(String connectionString)
+      throws IllegalArgumentException, MalformedURLException, SSLException {
+    if (connectionString == null || connectionString.isEmpty()) {
+      throw new IllegalArgumentException("Connection string cannot be null or empty");
+    }
+
+    // Connection string format: dgraph://[username:password@]host:port[?params]
+    if (!connectionString.startsWith("dgraph://")) {
+      throw new IllegalArgumentException("Invalid connection string: scheme must be 'dgraph'");
+    }
+    // Parse the URL (Use java.net.URL initially to validate the basic structure)
+    URL url;
+    try {
+      // Replace dgraph:// with http:// for proper URL parsing
+      url = new URL(connectionString.replace("dgraph://", "http://"));
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException("Failed to parse connection string: " + e.getMessage(), e);
+    }
+
+    // Extract host and port
+    String host = url.getHost();
+    int port = url.getPort();
+
+    if (host == null || host.isEmpty()) {
+      throw new IllegalArgumentException("Invalid connection string: hostname required");
+    }
+    if (port == -1) {
+      throw new IllegalArgumentException("Invalid connection string: port required");
+    }
+
+    // Extract username and password
+    String username = null;
+    String password = null;
+    if (url.getUserInfo() != null) {
+      String[] userInfo = url.getUserInfo().split(":", 2);
+      username = userInfo[0];
+      if (userInfo.length > 1) {
+        password = userInfo[1];
+      }
+    }
+
+    if (username != null && (password == null || password.isEmpty())) {
+      throw new IllegalArgumentException(
+          "Invalid connection string: password required when username is provided");
+    }
+
+    // Parse parameters into a Map (crazy that there's no built-in support for this in net.URL)
```

Comment:
Can we move the parsing code into another function?

Reply:
We can of course, but who else will use it?

Reply:
This function will be smaller and easier to read.